src/main/scala/org/antipathy/mvn_scalafmt/SourcesBuilder.scala

```diff
@@ -0,0 +1,54 @@
+package org.antipathy.mvn_scalafmt
+
+import java.io.File
+import java.util.{List => JList}
+
+import scala.jdk.CollectionConverters._
+
+import org.apache.maven.project.MavenProject
+
+/** Class for building a set of source dirs in the maven project */
+class SourcesBuilder(project: MavenProject) {
+
+  private val sources = Set.newBuilder[File]
+  private val build   = project.getBuild
+
+  private[mvn_scalafmt] def resultSet(): Set[File] =
+    try sources.result()
+    finally sources.clear()
+
+  def result(): JList[File] = resultSet().toList.asJava
+
+  def addMain(dirs: JList[File]): Unit =
+    if (!add(dirs)) {
+      appendPrimary(build.getSourceDirectory)
+      appendAlternative(project.getCompileSourceRoots, build.getOutputDirectory)
+    }
+
+  def addTest(dirs: JList[File]): Unit =
+    if (!add(dirs)) {
+      appendPrimary(build.getTestSourceDirectory)
+      appendAlternative(project.getTestCompileSourceRoots, build.getTestOutputDirectory)
+    }
+
+  private def add(dirs: JList[File]): Boolean = {
+    val ok = dirs != null && !dirs.isEmpty
+    if (ok) sources ++= dirs.asScala
+    ok
+  }
+
+  private def getCanonicalFile(dir: String): File = {
+    val relfile = new File(dir)
+    val absfile = if (relfile.isAbsolute) relfile else new File(project.getBasedir, dir)
+    absfile.getCanonicalFile
+  }
+
+  private def appendPrimary(dir: String): Unit =
+    sources += getCanonicalFile(dir + "/../scala")
+
+  private def appendAlternative(altDirs: JList[String], outputDir: String): Unit = {
+    val outputPath = getCanonicalFile(outputDir).toPath
+    sources ++= altDirs.asScala.map(getCanonicalFile).filter(!_.toPath.startsWith(outputPath))
```

Comment:
compare to #149: added the logic excluding generated files (those under the output dirs).